src/mlpack/methods/softmax_regression/softmax_regression.hpp

```diff
@@ -156,17 +156,23 @@ class SoftmaxRegression
    * Train the softmax regression with the given training data.
    *
    * @tparam OptimizerType Desired optimizer type.
+   * @tparam CallbackTypes Types of Callback Functions.
    * @param data Input data with each column as one example.
    * @param labels Labels associated with the feature data.
    * @param numClasses Number of classes for classification.
    * @param optimizer Desired optimizer.
    * @return Objective value of the final point.
+   * @param callbacks Callback function for ensmallen optimizer `OptimizerType`.
+   *      See https://www.ensmallen.org/docs.html#callback-documentation.
+   * 
+   * @return The final objective of the trained model (NaN or Inf on error)
    */
-  template<typename OptimizerType = ens::L_BFGS>
+  template<typename OptimizerType = ens::L_BFGS, typename... CallbackTypes>
   double Train(const arma::mat& data,
                const arma::Row<size_t>& labels,
                const size_t numClasses,
-               OptimizerType optimizer = OptimizerType());
+               OptimizerType optimizer = OptimizerType(),
+               CallbackTypes&&... callbacks);
```

Comment:
I don't think that we can add variadic parameters to a function that has default arguments.  You might have to split this into two functions.  Be sure to compile and test the changes locally before pushing, otherwise the builds will probably fail. :+1:

Reply:
@rcurtin  thanks

Reply:
About the build issue on macOS and Windows, I can see that this can be confusing as it builds with gcc but doesn't when using clang or MSVC. Anyway, if you take into account what @rcurtin said and take a look at the `SoftmaxRegression` constructor you might see, it's the same issue.

Also, please take another look at the solution you used here, as @rcurtin pointed out we want to split this into two functions:

  ```c++

template<typename OptimizerType = ens::L_BFGS, typename... CallbackTypes>
  double Train(const arma::mat& data,
               const arma::Row<size_t>& labels,
               const size_t numClasses,
               OptimizerType optimizer = OptimizerType());

template<typename OptimizerType, typename... CallbackTypes>
  double Train(const arma::mat& data,
               const arma::Row<size_t>& labels,
               const size_t numClasses,
               OptimizerType optimizer,
               CallbackTypes&&... callbacks);
```

That way we don't break reverse compatibility. Let me know if I should clarify anything.